text/0000-optimise-attr.md

```diff
@@ -0,0 +1,120 @@
+would result in larger code rather than a smaller one. It may also have effect on what instructions
+are selected to appear in the final binary.
+
+Note that `#[optimise(size)]` is a hint, rather than a hard requirement and compiler may still,
+while optimising, take decisions that increase function size compared to an entirely unoptimised
+result.
+
+Using this attribute is recommended when inspection of generated code reveals unnecessarily large
+function or functions, but use of `-O` is still preferable over `-C opt-level=s` or `-C
+opt-level=z`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `#[optimise(size)]` attribute applied to a function definition will instruct the optimisation
+engine to avoid applying optimisations that could result in a size increase and machine code
+generator to generate code that’s smaller rather than larger.
+
+Note that the `#[optimise(size)]` attribute is just a hint and is not guaranteed to result in any
+different or smaller code.
+
+Since `#[optimise(size)]` instructs optimisations to behave in a certain way, this means that this
+attribute has no effect when no optimisations are run (such as is the case when `-Copt-level=0`).
+Interaction of this attribute with the `-Copt-level=s` and `-Copt-level=z` flags is not specified
+and is left up to implementation to decide.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* Not all of the alternative codegen backends may be able to express such a request, hence the
+“this is an optimisation hint” note on the `#[optimise(size)]` attribute.
+    * As a fallback, this attribute may be implemented in terms of more specific optimisation hints
+      (such as `inline(never)`, the future `unroll(never)` etc).
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+Proposed is a very semantic solution (describes the desired result, instead of behaviour) to the
+problem of needing to sometimes inhibit some of the trade-off optimisations such as loop unrolling.
+
+Alternative, of course, would be to add attributes controlling such optimisations, such as
+`#[unroll(no)]` on top of a a loop statement. There’s already precedent for this in the `#[inline]`
+annotations.
+
+The author would like to argue that we should eventually have *both*, the `#[optimise(size)]` for
+people who look at generated code and decide that it is too large, and the targetted attributes for
+people who know *why* the code is too large.
+
+Furthermore, currently `optimise(size)` is able to do more than any possible combination of
+targetted attributes would be able to such as influencing the instruction selection or switch
+codegen strategy (jump table, if chain, etc.) This makes the attribute useful even in presence of
+all the targetted optimisation knobs we might have in the future.
+
+---
+
+Alternative: `optimize` (American English) instead of `optimise`… or both?
```

Comment:
Due to the precedence set by GCC and the fact that the vast majority of conversations I've read use american english spelling ("optimize", "optimizer") I'd prefer that over "optimise".

Reply:
For sure not both. As sad as this makes me (given that I prefer BrE), use of AmE is standard in the software industry, so we should stick with that.

Reply:
At the risk of sounding pedantic (or worse, boring), I feel I should point out that the OED considers 'ize' to be [perfectly acceptable and indeed preferred](https://blog.oxforddictionaries.com/2011/03/28/ize-or-ise/) BrE spelling.

Reply:
@jesskfullwood You always learn something useful; and today is such a day :)